homie-influx/src/main.rs

```diff
@@ -0,0 +1,388 @@
+                                device_id,
+                                node_id,
+                                property_id,
+                            )
+                            .await;
+                        }
+                    }
+                    _ => {
+                        log::info!("Event: {:?}", event);
+                    }
+                }
+            }
+        }
+    })
+}
+
+async fn send_property_value(
+    controller: &HomieController,
+    influx_db_client: &Client,
+    device_id: String,
+    node_id: String,
+    property_id: String,
+) {
+    if let Some(device) = controller.devices().get(&device_id) {
+        if let Some(node) = device.nodes.get(&node_id) {
+            if let Some(property) = node.properties.get(&property_id) {
+                if let Some(point) =
+                    point_for_property_value(device, node, property, SystemTime::now())
+                {
+                    // TODO: What should rp be?
+                    // TODO: Handle errors
+                    influx_db_client
+                        .write_point(point, INFLUXDB_PRECISION, None)
+                        .await
+                        .unwrap();
+                }
+            }
+        }
+    }
+}
+
+/// Convert the value of the given Homie property to an InfluxDB value of the appropriate type, if
+/// possible. Returns None if the datatype of the property is unknown, or there was an error parsing
+/// the value.
+fn influx_value_for_homie_property(property: &Property) -> Option<Value> {
+    let datatype = property.datatype?;
+    Some(match datatype {
+        Datatype::Integer => Value::Integer(property.value().ok()?),
+        Datatype::Float => Value::Float(property.value().ok()?),
+        Datatype::Boolean => Value::Boolean(property.value().ok()?),
+        _ => Value::String(property.value.to_owned()?),
+    })
+}
+
+/// Construct an InfluxDB `Point` corresponding to the given Homie property value update.
+fn point_for_property_value(
+    device: &Device,
+    node: &Node,
+    property: &Property,
+    timestamp: SystemTime,
+) -> Option<Point> {
+    let datatype = property.datatype?;
+    let value = influx_value_for_homie_property(property)?;
+
+    let mut point = Point::new(&datatype.to_string())
+        .add_timestamp(
+            timestamp
+                .duration_since(SystemTime::UNIX_EPOCH)
+                .unwrap()
+                .as_millis() as i64,
+        )
+        .add_field("value", value)
+        .add_tag("device_id", Value::String(device.id.to_owned()))
+        .add_tag("node_id", Value::String(node.id.to_owned()))
+        .add_tag("property_id", Value::String(property.id.to_owned()));
+    if let Some(device_name) = device.name.to_owned() {
+        point = point.add_field("device_name", Value::String(device_name));
+    }
+    if let Some(node_name) = node.name.to_owned() {
+        point = point.add_field("node_name", Value::String(node_name));
```

Comment:
I think that these should be tags as well: If I want to plot a graph of all fridge temperature measurements, I want node_name and property_name to be indexed (I still haven't worked out how the indexes work in-memory, or how indexes for different tags are combined, so my instincts might be wrong).

We're not expecting either of these things to be high cardinality, so I think tag is probably more appropriate: https://docs.influxdata.com/influxdb/v1.8/concepts/schema_and_data_layout/

Reply:
When I say "high cardinality", even a low-spec influxdb box will be fine with a series cardinality of 100,000 according to https://docs.influxdata.com/influxdb/v1.8/guides/hardware_sizing/.

See also https://docs.influxdata.com/influxdb/v1.8/concepts/glossary/#series-cardinality - seems like serieses are grouped by `(database, measurement, tag set, field key)` when determining cardinality. We can probably infer something about how the indexes are structured from this. I would assume that each series is recorded as a column, and that you calculate the tag sets that you're looking for first, and then combine the every series that matches your filters into a single series before rendering?

Reply:
I suppose the most compelling reason to make it a tag is that if I patch this field to be a tag then I can do group by `node_name` and alias by `[[tag_node_name]]` and get the following graph with names straight away:

<img width="1017" alt="Screenshot 2020-11-09 at 15 28 27" src="https://user-images.githubusercontent.com/254647/98561186-7d771500-22a0-11eb-8c8f-191b341881e7.png">


 (I'm not sure whether `explore` urls are portable, but this works on my local grafana instance and might be useful to you as well? http://localhost:3000/explore?orgId=1&left=%5B%22now-5m%22,%22now%22,%22InfluxDB%22,%7B%22datasource%22:%22InfluxDB%22,%22policy%22:%22default%22,%22resultFormat%22:%22time_series%22,%22orderByTime%22:%22ASC%22,%22tags%22:%5B%7B%22key%22:%22property_id%22,%22operator%22:%22%3D%22,%22value%22:%22temperature%22%7D%5D,%22groupBy%22:%5B%7B%22type%22:%22time%22,%22params%22:%5B%22$__interval%22%5D%7D,%7B%22type%22:%22tag%22,%22params%22:%5B%22node_name%22%5D%7D%5D,%22select%22:%5B%5B%7B%22type%22:%22field%22,%22params%22:%5B%22value%22%5D%7D,%7B%22type%22:%22mean%22,%22params%22:%5B%5D%7D%5D%5D,%22measurement%22:%22float%22,%22alias%22:%22%5B%5Btag_node_name%5D%5D%22,%22hide%22:false%7D,%7B%22ui%22:%5Btrue,true,true,%22none%22%5D%7D%5D )

Reply:
Okay, done.